src/shared/pkcs11-util.c

```diff
@@ -632,96 +668,276 @@ int pkcs11_token_find_private_key(
+                { CKA_KEY_TYPE, (void *)&shared_secret_type, sizeof(shared_secret_type) },
+                { CKA_VALUE_LEN, (void *)&shared_secret_len, sizeof(shared_secret_len) },
+                { CKA_SENSITIVE, (void *)&no, sizeof(no) },
+                { CKA_EXTRACTABLE, (void *)&yes, sizeof(yes) },
+                { CKA_ENCRYPT, (void *)&yes, sizeof(yes) },
+                { CKA_DECRYPT, (void *)&yes, sizeof(yes) }
+        };
+        CK_ATTRIBUTE shared_secret_attr = {
+                .type = CKA_VALUE,
+        };
+        CK_MECHANISM mechanism = {
+                .mechanism = CKM_ECDH1_DERIVE,
+                .pParameter = &params,
+                .ulParameterLen = sizeof(CK_ECDH1_DERIVE_PARAMS)
+        };
+
+        if (EVP_PKEY_base_id(peer_key) != EVP_PKEY_EC)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "ECDH peer key is not a EC key");
+
+        /* Curves like secp521r1 will return 521 for EVP_PKEY_bits call,
+         * We need round it up to a size that can hold the generated shared secret */
+        shared_secret_len = (EVP_PKEY_bits(peer_key) + 7) / 8;
+
+        if (!ret_shared_secret)
+                *ret_shared_secret_len = shared_secret_len;
+
+#if OPENSSL_VERSION_NUMBER >= 0x30000000
+        if (EVP_PKEY_set_utf8_string_param(peer_key,
+                                          OSSL_PKEY_PARAM_EC_POINT_CONVERSION_FORMAT,
+                                          OSSL_PKEY_EC_POINT_CONVERSION_FORMAT_UNCOMPRESSED) != 1)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO),
+                                                       "Failed to set PKEY parameter \""
+                                                       OSSL_PKEY_PARAM_EC_POINT_CONVERSION_FORMAT "\" to \""
+                                                       OSSL_PKEY_EC_POINT_CONVERSION_FORMAT_UNCOMPRESSED "\"");
+
+        public_data_len = EVP_PKEY_get1_encoded_public_key(peer_key, &buffer);
+        if (public_data_len == 0 || buffer == NULL)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to get encoded public key from PKEY");
+#else
+        EC_KEY *ec_key = EVP_PKEY_get0_EC_KEY(peer_key);
+        if (!ec_key)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Peer key doesn't have associated EC_KEY");
+
+        if (EC_KEY_check_key(ec_key) != 1)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "EC_KEY associated with PKEY is not valid");
+
+        public_data_len = EC_POINT_point2oct(EC_KEY_get0_group(ec_key),
+                                             EC_KEY_get0_public_key(ec_key),
+                                             POINT_CONVERSION_UNCOMPRESSED,
+                                             NULL, 0, NULL);
+        if (public_data_len == 0)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to determine public data size");
+
+        buffer = new(uint8_t, public_data_len);
```

Comment:
this is just a generic "blob" to us, not actually an array, hence should be `void*` rather than `uint8_t*`, and hence be allocated with a generic malloc(), not a type-safe new()